cpp/dolfinx/common/array_2d.h

```diff
@@ -0,0 +1,187 @@
+class array_2d
+{
+public:
+  /// \cond DO_NOT_DOCUMENT
+  using value_type = T;
+  using allocator_type = Allocator;
+  using size_type = typename std::vector<T, Allocator>::size_type;
+  using reference = typename std::vector<T, Allocator>::reference;
+  using const_reference = typename std::vector<T, Allocator>::const_reference;
+  using pointer = typename std::vector<T, Allocator>::pointer;
+  using iterator = typename std::vector<T, Allocator>::iterator;
+  using const_iterator = typename std::vector<T, Allocator>::const_iterator;
+  /// \endcond
+
+  /// Constructs an empty two dimensional array.
+  array_2d() : _rows(0), _cols(0) { _storage = std::vector<T, Allocator>(); }
+
+  /// Constructs a two dimensional array with size = rows * cols, and initialize
+  /// elements with value value.
+  array_2d(size_type rows, size_type cols, value_type value = T(),
+           const Allocator& alloc = Allocator())
+      : _rows(rows), _cols(cols)
+  {
+    _storage = std::vector<T, Allocator>(_rows * _cols, value, alloc);
+  }
+
+  /// Constructs a two dimensional array with size = rows * cols, and initialize
+  /// elements with value value.
+  template <class Iterator>
+  array_2d(const std::vector<size_t>& shape, Iterator first, Iterator last)
+  {
+    assert(shape.size() == 2);
+    _rows = shape[0];
+    _cols = shape[1];
+    _storage.resize(_rows * _cols);
+    std::copy(first, last, _storage.begin());
+  }
+
+  /// Constructs a two dimensional array with the copy of the contents of other
+  /// two dimensional container.
+  template <class Container,
+            typename
+            = typename std::enable_if<is_2d_container<Container>::value>::type>
+  array_2d(Container& array)
+  {
+    _rows = array.rows();
+    _cols = array.cols();
+    _storage.resize(array.size());
+    std::copy(array.data(), array.data() + array.size(), _storage.begin());
+  }
+
+  /// Constructs a two dimensional array with the contents of the initializer
+  /// list init.
+  array_2d(size_type rows, size_type columns, std::initializer_list<T> list,
+           const Allocator& alloc = Allocator())
+      : _rows(rows), _cols(columns)
+  {
+    if (_rows * _cols != list.size())
+      throw std::runtime_error("Dimension mismatch");
+    _storage = std::vector<T, Allocator>(list, alloc);
+  }
+
+  /// Returns a reference to the element at specified location (i, j).
+  /// No bounds checking is performed.
+  reference operator()(size_type i, size_type j)
+  {
+    return _storage[i * _cols + j];
+  }
+
+  /// Returns a const reference to the element at specified location (i, j).
+  /// No bounds checking is performed.
+  const_reference operator()(size_type i, size_type j) const
+  {
+    return _storage[i * _cols + j];
+  }
+
+  /// Returns a reference to the row at specified location (i).
+  tcb::span<value_type> row(int i) const
+  {
+    size_type offset = i * _cols;
+    return tcb::span<value_type>(pointer(&_storage[0] + offset), _cols);
```

Comment:
since it's not possible to convert a stl iterator to a pointer, I used instead;

```c++
return tcb::span<const value_type>(std::next(_storage.data(), offset), _cols);
```